:seedling: Adds the ability to add a custom Auth Button

gowthamidommety-okta opened the pull request:
This allows users to add a custom Auth Button via new properties when initializing the sign in widget.

Resolves: OKTA-119408

mauriciocastillosilva-okta commented on src/PrimaryAuthController.js:
```diff
@@ -14,11 +14,12 @@ define([
   'util/BaseLoginController'
 ],
-function (Okta, PrimaryAuthForm, SocialAuth, PrimaryAuthModel, Util, BaseLoginController) {
+function (Okta, PrimaryAuthForm, SocialAuth, AuthButton, PrimaryAuthModel, Util, BaseLoginController) {
```
we usually have the same name here as in the 'define' block. So this should be AddAuthButton

gowthamidommety-okta replied:
Makes a lot of sense. Initially I wanted to refactor Social Auth to be more generic, but SocialAuth extending CustomButtons is better

mauriciocastillosilva-okta commented on src/PrimaryAuthController.js:
```diff
@@ -141,9 +142,14 @@ function (Okta, PrimaryAuthForm, SocialAuth, PrimaryAuthModel, Util, BaseLoginCo
         this.add(SocialAuth, {prepend: options.settings.get('socialAuthPositionTop')});
       }
+      if (addButtonSettings && addButtonSettings.visibleWhen) {
+        this.add(AuthButton.render(addButtonSettings));
```
we should add an AuthButton, not call a render method. This should be
```
this.add(AuthButton);
```

gowthamidommety-okta replied:
Refactoring code from discussion we had. So this won't be an issue

mauriciocastillosilva-okta commented on src/models/Settings.js:
```diff
@@ -76,6 +76,11 @@ function (Okta, Q, Errors, BrowserFeatures, Util, Logger, config) {
       'features.deviceFingerprinting': ['boolean', false, false],
       'features.hideSignOutLinkInMFA' : ['boolean', false, false],
       'features.customExpiredPassword': ['boolean', true, false],
+      'features.addButton': ['object', false, false],
```
not sure you need this one, since you are explicitly defining the contents of it in the next few lines

mauriciocastillosilva-okta commented on src/views/primary-auth/AddAuthButton.js:
```diff
@@ -0,0 +1,49 @@
+          <span class="auth-divider-text">{{text}}</span>\
+        </div>');
+
+    var render = function(buttonSettings) {
```
no need for this, just return the View. 
If you need to access the settings, you should be able to though the model ```this.model.get('features.addButton.text')``` and so on

gowthamidommety-okta replied:
Refactoring code from discussion we had. So this won't be an issue

mauriciocastillosilva-okta commented on src/PrimaryAuthController.js:
```diff
@@ -141,9 +142,14 @@ function (Okta, PrimaryAuthForm, SocialAuth, PrimaryAuthModel, Util, BaseLoginCo
       if (options.settings.get('socialAuthConfigured')) {
         this.add(SocialAuth, {prepend: options.settings.get('socialAuthPositionTop')});
       }
+      if (addButtonSettings && addButtonSettings.visibleWhen) {
```
we need to evaluate if ```visibleWhen``` is a function or not. If it is, we need to call that function.

mauriciocastillosilva-okta replied:
Main concern with that is that we have a divider, that will show up above the button. If we support an array then how do we know if we add the divider or not? Or if we support adding them in different pages? It seems to me we could do that for v2.

mauriciocastillosilva-okta replied:
Actually, thinking about it more, we should always add the divider, and we can support an array of buttons.

mauriciocastillosilva-okta commented on test/unit/spec/PrimaryAuth_spec.js:
```diff
@@ -1808,4 +1824,41 @@ function (_, $, Q, OktaAuth, LoginUtil, Okta, Util, AuthContainer, PrimaryAuthFo
   });
 
+  Expect.describe('Additional Auth Button', function () {
+    itp('does not show the divider or buttons if addButton is not set', function () {
```
sp: ```divider and button```

gowthamidommety-okta replied:
Will change

mauriciocastillosilva-okta commented on test/unit/spec/PrimaryAuth_spec.js:
```diff
@@ -1808,4 +1824,41 @@ function (_, $, Q, OktaAuth, LoginUtil, Okta, Util, AuthContainer, PrimaryAuthFo
+        expect(test.form.additionalAuthButton().length).toBe(0);
+      });
+    });
+    itp('shows divider and buttons if properties passed with visibleWhen true', function () {
```
sp: singular ```button```

gowthamidommety-okta replied:
Will change

mauriciocastillosilva-okta commented on test/unit/spec/PrimaryAuth_spec.js:
```diff
@@ -1808,4 +1824,41 @@ function (_, $, Q, OktaAuth, LoginUtil, Okta, Util, AuthContainer, PrimaryAuthFo
+        expect(test.form.additionalAuthButton().length).toBe(0);
+      });
+    });
+    itp('sets text with property passed', function () {
```
one more test will be missing for when ```visibleWhen``` is a function

gowthamidommety-okta commented:
<img width="458" alt="customauthbutton" src="https://cloud.githubusercontent.com/assets/26014318/24559481/b0c542dc-15f3-11e7-87b7-6cff29400e6c.png">

gowthamidommety-okta commented on src/views/primary-auth/AddAuthButton.js:
```diff
@@ -0,0 +1,49 @@
+    var render = function(buttonSettings) {
+        return Okta.View.extend({
+
+            children: function () {
```
Thanks! Will change

gowthamidommety-okta commented on README.md:
```diff
@@ -957,6 +957,24 @@ features: {
 
+- **features.addButton** - Adds a custom Auth button configurable by the properties below:
+
+    - `text` - String to set button text
```
Added a screenshot below. This is only for the Primary Auth page. I don't think we can enable this for any other page.

gowthamidommety-okta commented:
Pushed new changes. Please ignore tests for now

mauriciocastillosilva-okta commented on src/PrimaryAuthController.js:
```diff
@@ -141,8 +141,9 @@ function (Okta, PrimaryAuthForm, SocialAuth, PrimaryAuthModel, Util, BaseLoginCo
 
       // Add SocialAuth view only when the idps are configured. If configured, 'socialAuthPositionTop'
```
you could move this comment to Settings.

gowthamidommety-okta replied:
Yea, i didn't see that. It doesn't fit in settings either. I will get rid of it.

mauriciocastillosilva-okta commented on src/models/Settings.js:
```diff
@@ -238,6 +247,57 @@ function (Okta, Q, Errors, BrowserFeatures, Util, Logger, config) {
+        fn: function (customButtons) {
+          _.each(customButtons, function(button) {
+            //add custom auth class name
+            button['className'] = button['className'] + ' default-custom-button';
```
I think we should also get rid of any other properties that are sent to us. Basically use a whitelist of what we need. Something like
```
_.pick(button, 'click', 'className', 'title')
```

gowthamidommety-okta replied:
I don't think we can.. Social Auth needs options (properties) sent to the call back.

rchild-okta replied:
Do we need both the idp & customButtons derived properties? Can we just combine this in one? Also, agree, we need to whitelist the properties that are sent to us. Similar to the other comment - but create new objects here, don't modify the existing one they pass in. Doing this, we can just pick and choose (like @mauriciocastillosilva-okta suggested) what we want to take from it).

mauriciocastillosilva-okta commented on src/models/Settings.js:
```diff
@@ -238,6 +247,57 @@ function (Okta, Q, Errors, BrowserFeatures, Util, Logger, config) {
+        deps: ['customButtonsOptionsSet', 'configuredIdpsOptionsSet', 'socialAuthConfigured'],
+        fn: function(customButtonsOptionsSet, configuredIdpsOptionsSet, socialAuthConfigured) {
+          var buttonArray = [];
+          if(socialAuthConfigured) {
```
you could also just check the length of ```configuredIdpsOptionsSet``` instead of ```socialAuthConfigured```. And if ```socialAuthConfigured``` is no longer used, you could even get rid of it.

gowthamidommety-okta replied:
socialAuthConfigured is used in socialAuthPositionTop. Since we use it in more than one place , I'm making use of it here.

rchild-okta replied:
Yeah, just reiterating - since the customButtonOptions and idpOptions are only really used for this (I'm guessing), let's just pack it all in one derived property.

rchild-okta commented on README.md:
````diff
@@ -918,6 +918,23 @@ Options for the [OpenId Connect](http://developer.okta.com/docs/api/resources/oi
       authScheme: 'OAUTH2'
     }
     ```
+- **customButtons** - Adds an array of custom Auth buttons configurable by the properties below:
````
This shouldn't be in the OIDC section, should it? I.e. it could be part of socialAuth, but you could also have a custom button outside of that flow right? Probably just create a new section (like helpLinks before this).

Also, in the style of helplinks, should reword to something like:

You can add custom buttons underneath the login form on the primary auth page by setting the following config option. If you'd like to change the divider text, use the `i18n` config option.

rchild-okta commented on README.md:
````diff
@@ -918,6 +918,23 @@ Options for the [OpenId Connect](http://developer.okta.com/docs/api/resources/oi
     ```
+- **customButtons** - Adds an array of custom Auth buttons configurable by the properties below:
+
+    - `title` - String to set button text
````
We *mostly* don't word things like this. Actually, this is a good reminder, @mystiberry-okta can you check the text in this PR? This is going to show up in our widget README (and later on, on d.o.c.).

Better perhaps to say something like "the button text"

rchild-okta commented on README.md:
```diff
@@ -918,6 +918,23 @@ Options for the [OpenId Connect](http://developer.okta.com/docs/api/resources/oi
+- **customButtons** - Adds an array of custom Auth buttons configurable by the properties below:
+
+    - `title` - String to set button text
+    - `className` - Sets a class on the button
```
Optional class that can be added to the button

rchild-okta commented on README.md:
```diff
@@ -918,6 +918,23 @@ Options for the [OpenId Connect](http://developer.okta.com/docs/api/resources/oi
+
+    - `title` - String to set button text
+    - `className` - Sets a class on the button
+    - `dataAttr` - Sets data-se attribute on the button
```
I don't think we want to expose this publicly - data-se is our own internal data attr for selenium testing. They'd never need to set this.

rchild-okta commented on README.md:
```diff
@@ -918,6 +918,23 @@ Options for the [OpenId Connect](http://developer.okta.com/docs/api/resources/oi
+    - `title` - String to set button text
+    - `className` - Sets a class on the button
+    - `dataAttr` - Sets data-se attribute on the button
+    - `click` - Callback function that is executed when button is clicked
```
Function that is called when the button is clicked

rchild-okta commented on README.md:
```diff
@@ -918,6 +918,23 @@ Options for the [OpenId Connect](http://developer.okta.com/docs/api/resources/oi
+      className: 'btn-customAuth',
+      dataAttr: 'custom-data',
+      click: function() {
+        //custom callback
```
nit: space between // and custom 

I do think it's more useful to add a real example, i.e. you could show the code to redirect to a different url (to give more context on what they'll actually be doing. With a comment before it saying that this is an example showing a redirect to bla.

rchild-okta commented on src/PrimaryAuthController.js:
```diff
@@ -139,10 +139,11 @@ function (Okta, PrimaryAuthForm, SocialAuth, PrimaryAuthModel, Util, BaseLoginCo
-        this.add(SocialAuth, {prepend: options.settings.get('socialAuthPositionTop')});
+      // If social auth is configured, 'socialAuthPositionTop' will determine
+      // the order in which the social auth and primary auth are shown on the screen.
+
```
nit: remove this newline (i.e. comment should be attached to the code it's talking about)

rchild-okta commented on src/models/Settings.js:
```diff
@@ -238,6 +241,57 @@ function (Okta, Q, Errors, BrowserFeatures, Util, Logger, config) {
+      configuredIdpsOptionsSet: {
+        deps: ['configuredSocialIdps'],
+        fn: function (configuredSocialIdps) {
+          _.each(configuredSocialIdps, function(idp) {
```
In general, we shouldn't be modifying the object they pass in - low risk, but it's possible they are doing something with it as well and changing the properties on the object could affect the developer's code.

```javascript
return _.map(configuredSocialIdps, function (idp) {
  var type = idp.type.toLowerCase();
  return {
    // the properties + idp properties
    // alternatively, clone the idp and then add these properties to it
  };
});
```

And a nit - no need to use `idp['dataAttr']`, use `idp.dataAttr`

rchild-okta commented on src/views/primary-auth/CustomButtons.js:
```diff
@@ -56,21 +55,19 @@ define([
-          attr = 'social-auth-' + type + '-button';
-
+      if(options.isSocialAuth) {
+        options.click =  function(e) {
```
Why not just do the click function in settings.js? In general, better to have things look the same - in this case, SocialAuth is using the generic CustomButtons class, rather than having these edge cases built into this generic view.

gowthamidommety-okta replied:
It's difficult to have this in Settings.js because the callback references other classes as well. See OAuth2Util.

rchild-okta commented on src/views/primary-auth/CustomButtons.js:
```diff
@@ -27,16 +27,15 @@ define([
-          idProviders = this.settings.get('configuredSocialIdps'),
-          divider = dividerTpl({text: Okta.loc('socialauth.divider.text', 'login')});
+        buttons = this.settings.get('configuredButtons'),
+        divider = dividerTpl({text: Okta.loc('authbutton.divider.text', 'login')});
```
Hmm, I am a little hesitant about this change. It is a breaking change - should think about this more (i.e. if someone updates the widget, is customizing this string, and is using socialAuth, this will result in a change of behavior for them).

gowthamidommety-okta replied:
Spoke to @mauriciocastillosilva-okta about this. I'll change it back to use socialauth.divider.text.

rchild-okta commented on test/unit/spec/PrimaryAuth_spec.js:
```diff
@@ -127,6 +127,22 @@ function (_, $, Q, OktaAuth, LoginUtil, Okta, Util, AuthContainer, PrimaryAuthFo
     });
   }
 
+  function setupAdditionalAuthButton() {
```
General: Since we have e2e tests, might be nice to put a new one in there as well.

mauriciocastillosilva-okta replied:
Do this qualifies for an e2e test? At the end we are just calling the callback they sent us, nothing else happens in the widget (and we don't get a token back).

gowthamidommety-okta commented:
Made more changes. @mauriciocastillosilva-okta @rchild-okta Please review. Some notes:

**Settings.js:** 
- Removed socialAuthConfigured and its dependancies for simplification.
- Did not remove socialAuthPositionTop as it is used in PrimaryAuthController , PrimaryAuthForm and the CustomButtons file.
- Consolidated some of the other new dependancies

**PrimaryAuthForm.js:**
- removed the check for socialAuthConfigured as socialAuthPositionTop is derived from it anyway

**tests**
- Added some tests to check undefined

mauriciocastillosilva-okta commented on src/models/Settings.js:
```diff
@@ -223,19 +227,65 @@ function (Okta, Q, Errors, BrowserFeatures, Util, Logger, config) {
+          }
+          var buttonArray = [];
+          if(!_.isEmpty(socialAuthButtons)) {
+            buttonArray = buttonArray.concat(socialAuthButtons);
```
can we, instead of concatenating 2 arrays here, use the same array since the beginning and skip the concat part? At the end we just need to add buttons to the same array and we can return that
```
var buttons = [];
// add social buttons to _buttons_
// add custom buttons to _buttons_
return buttons;
```

mauriciocastillosilva-okta commented on test/unit/spec/PrimaryAuth_spec.js:
```diff
@@ -1808,4 +1824,105 @@ function (_, $, Q, OktaAuth, LoginUtil, Okta, Util, AuthContainer, PrimaryAuthFo
   });
 
+  Expect.describe('Additional Auth Button', function () {
+    itp('does not show the divider and buttons if addButton is not set', function () {
```
should read ```...  buttons if settings.customButtons is not set```

mauriciocastillosilva-okta commented on test/unit/spec/PrimaryAuth_spec.js:
```diff
@@ -1808,4 +1824,105 @@ function (_, $, Q, OktaAuth, LoginUtil, Okta, Util, AuthContainer, PrimaryAuthFo
+        expect(test.form.additionalAuthButton().length).toBe(0);
+      });
+    });
+    itp('show the divider and buttons if addButton set', function () {
```
should read ```...  buttons if settings.customButtons is not empty```

gowthamidommety-okta commented:
Please review again!

mauriciocastillosilva-okta commented on src/models/Settings.js:
```diff
@@ -223,19 +227,55 @@ function (Okta, Q, Errors, BrowserFeatures, Util, Logger, config) {
+      socialAuthPositionTop: {
+        deps: ['configuredSocialIdps', 'idpDisplay'],
+        fn: function (configuredSocialIdps, idpDisplay) {
+          return !!(!_.isEmpty(configuredSocialIdps) && idpDisplay.toUpperCase() === 'PRIMARY');
```
nit: not sure if you need !! at the beginning. Both conditions return a boolean...

rchild-okta commented on README.md:
```diff
@@ -776,6 +776,28 @@ helpLinks: {
+  className: 'btn-customAuth',
+  click: function() {
+   // clicking on the button navigates to another page
+   window.location.href = 'http://www.example.com'
```
nit: spacing and semicolon at the end

rchild-okta commented on README.md:
```diff
@@ -776,6 +776,28 @@ helpLinks: {
 
 - **helpLinks.custom** - Array of custom link objects `{text, href}` that will be added to the *"Need help signing in?"* section.
 
+## Adding Custom Buttons
```
Add to menu up top

rchild-okta commented on src/models/Settings.js:
```diff
@@ -223,19 +227,55 @@ function (Okta, Q, Errors, BrowserFeatures, Util, Logger, config) {
-        deps: ['socialAuthConfigured', 'idpDisplay'],
-        fn: function (socialAuthConfigured, idpDisplay) {
-          return !!(socialAuthConfigured && idpDisplay.toUpperCase() === 'PRIMARY');
+      //merges social auth and custom buttons into one array
```
nit: space after comment //

rchild-okta commented on src/models/Settings.js:
```diff
@@ -223,19 +227,55 @@ function (Okta, Q, Errors, BrowserFeatures, Util, Logger, config) {
+            };
+            buttonArray.push(socialAuthButton);
+          });
+          if(_.isArray(customButtons)) {
```
Do we actually need the check? I.e. I'm guessing our validator should handle it, and the default value is an empty array.

gowthamidommety-okta replied:
When i set customButtons as any string in .widgetrc, it renders a lot of buttons. I'm not sure if this is  a valid use case, but i put in a _.isArray check to stop it. Spoke to Mau about it too, and we decided there is no harm in leaving it in
